### DISCUSSION TITLE
Time x Space tutorial with basis representation and FPCA ?

### DISCUSSION THREAD

**Discussion by mat-ej** [GAA-UAM/scikit-fda#569] [Q&A]
Hi Scikit-Fda team, 

First, let me start by saying that this is one of the best packages I have ever stumbled upon. Your tutorials are top notch introduction into functional data analysis. I love the way it is similar and works beautifully with original sci kit learn. 

What I am missing is a tutorial or perhaps just a short guide on how to work with Time x (Latitude, Longitude) data. I have found the tutorial for dealing with time and I have found tutorial for dealing with surfaces.

[https://fda.readthedocs.io/en/latest/auto_examples/plot_composition.html](https://fda.readthedocs.io/en/latest/auto_examples/plot_composition.html)

The tutorials are very good, but I am having trouble understanding how to construct basis representation for 
(Latitude x Longitude x Time) data, how can I perform FPCA in that case etc. 

Is there such a tutorial or perhaps a slack community where I might get some pointers? : ) 

Thanks in advance for your help and such an awesome package. 

Have a great week.

**Comment by vnmabus** [comment]
> First, let me start by saying that this is one of the best packages I have ever stumbled upon. Your tutorials are top notch introduction into functional data analysis. I love the way it is similar and works beautifully with original scikit learn.

Thank you! Making our work compatible with scikit-learn is one of our core principles, as it allows familiarity and code reuse, and simplifies some workflows in which both functional and multivariate data appear.

> What I am missing is a tutorial or perhaps just a short guide on how to work with Time x (Latitude, Longitude) data. I have found the tutorial for dealing with time and I have found tutorial for dealing with surfaces.
> 
> https://fda.readthedocs.io/en/latest/auto_examples/plot_composition.html

As functional data analysis is a large topic, we have focused first in the 1D (curves) case, in both the code and documentation. I agree that we should attempt to improve that situation. The state of the package (develop branch) in the ND case is the following:
- Functions of several variables discretized in a complete grid can be represented as a [`FDataGrid`](https://fda.readthedocs.io/en/latest/modules/autosummary/skfda.representation.grid.FDataGrid.html).
- Some bases, such as [`TensorBasis`](https://fda.readthedocs.io/en/latest/modules/autosummary/skfda.representation.basis.TensorBasis.html) and [`FiniteElementBasis`](https://fda.readthedocs.io/en/latest/modules/autosummary/skfda.representation.basis.FiniteElementBasis.html) can be employed to create a basis for functions of several variables.
- The conversion from `FDataGrid` to [`FDatabasis`](https://fda.readthedocs.io/en/latest/modules/autosummary/skfda.representation.basis.FDataBasis.html) should work (but may be slow or use much memory).
- Many of the methods that work with `FDataGrid` do not attempt to work with functions of several variables. A notable exception are methods based on distances, which will work if the distance is defined for such functions (e.g. the default L2 distance).
- As a design principle, we try to make the methods work in general if it is not too difficult. However, if it slows development too much we focus in the 1D case. Methods that do not support several variables should raise an exception when that type of data is provided (but as we do not focus for now in that case, it is probable that some failed to do so and produce bogus results instead. This should be treated as a bug).
- Methods that use `FDataBasis` tend to work also with basis for functions of several variables, as they are usually based on the inner product, which is implemented for that case. However, they MAY be slow or use much memory. An example of these methods would be FPCA for `FDataBasis`.
- Unfortunately, we have not yet implemented [`LinearDifferentialOperator`](https://fda.readthedocs.io/en/latest/modules/misc/autosummary/skfda.misc.operators.LinearDifferentialOperator.html) for the ND case, so smoothing based on the derivatives will not work.

> The tutorials are very good, but I am having trouble understanding how to construct basis representation for (Latitude x Longitude x Time) data, how can I perform FPCA in that case etc.

- If your data has been measured in a complete grid Latitude x Longitude x Time you can represent it as a `FDataGrid`.
- Then, you can create a [`TensorBasis`](https://fda.readthedocs.io/en/latest/modules/autosummary/skfda.representation.basis.TensorBasis.html) passing to it the individual desired 1D bases for the Latitude, Longitude and Time.
- You can convert your original discretized data to `FDataBasis` using either the `to_basis` method or the `BasisSmoother` class.
- You can then use the [`FPCA`](https://fda.readthedocs.io/en/latest/modules/preprocessing/autosummary/skfda.preprocessing.dim_reduction.FPCA.html) class to perform FPCA in basis form.

> Is there such a tutorial or perhaps a slack community where I might get some pointers? : )

As I said, we should improve the tutorial including topics such as registration, classification or regression, and probably also a section on functions of several variables.
We do not have a Slack (nor Discord nor Matrix) community. I agree that it would be an amazing resource to have, but we do not have currently the time to manage one. We have however the [Github discussions](https://github.com/GAA-UAM/scikit-fda/discussions) open, so that users of the package (or functional data analysis researchers/practitioners, even if they do not use it) can ask and answer questions or comment in related topics.

> Thanks in advance for your help and such an awesome package.

Thank you again! I hope that my answers helped you.

**Reply by mat-ej** [reply]
Hi Carlos, 

Thanks for making this into a discussion. 

As I have mentioned I am having issues understanding the **Time x Latitude x Longitude** basis and performing FPCA on such a representation. 

My data is measurements across Time at different stations placed at Lat x Lon coordinates 
![github_samples](https://github.com/GAA-UAM/scikit-fda/assets/10036980/dff0fb90-3d17-4624-ba83-08e95d0c6eed)

where mean trace is 
![github_mean](https://github.com/GAA-UAM/scikit-fda/assets/10036980/002e3754-e98f-4072-a4f0-0396a94b3db6)

I am able to succesfully use the TensorBasis: 

```
n_basis_functions = 10
space_basis = TensorBasis([
    BSplineBasis(  # Longitude
        n_basis=n_basis_functions,
    ),
    BSplineBasis(  # Latitude
        n_basis=n_basis_functions,
    ),
])

```
to smooth the "Surface"

![space_basis](https://github.com/GAA-UAM/scikit-fda/assets/10036980/d3c36c7d-0634-49ce-953f-2b8dc42cdc35)

I am however having trouble with the Time dimension. Perhaps it is due to the nature of the problem. 
```
time_basis = FourierBasis(n_basis = 50)
```
will oversmooth:

![time_basis](https://github.com/GAA-UAM/scikit-fda/assets/10036980/4d87c69e-55df-47cf-8275-6336e345498a)


And basically the Timeseries is so long that no number of basis functions will make the smoothed series resemble the original data.

**Q:** Is there a way for me to use the TensorBasis to represent all dimensions? Something like the following snippet ? 

```
TensorBasis([
    BSplineBasis(  # X axis
        n_basis= 10,
    ),
    BSplineBasis(  # Y axis
        n_basis= 10,
    ),
    FourierBasis( # Time
        n_basis = 100
    )
])
```

(Currently I only represent the surface in a functional form and then do FPCA without considering the time, would considering the time bring me any advantage ?)

Thanks so much for help : )

**Reply by vnmabus** [reply]
> My data is measurements across Time at different stations placed at Lat x Lon coordinates

Then I guess that each station has one curve and two scalar values associated with it, like in Canadian Weather. Sorry, because I first thought that your data consisted on 3D curves (Time x Latitude x Longitude) but that is apparently not the case.

This type of data, consisting in both functional and scalar attributes is called mixed data. We can currently represent it in a Pandas dataframe, as `FDataGrid` and `FDataBasis` can be used as Pandas columns. However, we do not have currently much functionality that is able to work with this representation (right now, just [linear regression with scalar response](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.LinearRegression.html#skfda.ml.regression.LinearRegression)). That is why, for example, in [this example](https://fda.readthedocs.io/en/latest/auto_examples/full_examples/plot_aemet_unsupervised.html#sphx-glr-auto-examples-full-examples-plot-aemet-unsupervised-py) we use only the functional part for the analysis, and the latitude and longitude are just for visualizing the results in a map.

> I am however having trouble with the Time dimension.

Have you tried the `BSpineBasis` for time? It does not have the "oscillations" of Fourier. It could probably approach your data better, at least with a high number of elements of the basis. That said, it is probable that the time dimension here is not very informative, as in the plot the curves seemed very similar (but I cannot tell for sure, as I do not know your problem).

> **Q:** Is there a way for me to use the TensorBasis to represent all dimensions? Something like the following snippet ?
> 
> ```
> TensorBasis([
>     BSplineBasis(  # X axis
>         n_basis= 10,
>     ),
>     BSplineBasis(  # Y axis
>         n_basis= 10,
>     ),
>     FourierBasis( # Time
>         n_basis = 100
>     )
> ])
> ```

The problem is that in that case you would have just one function representing all the measurements for all the stations, wouldn't you? I do not know if there is something that can be done with such a representation, but that probably is outside the focus of this package. You certainly cannot do statistics with just one sample.

Moreover, in case that you wanted to do that, we would need to find the appropriate basis coefficients, and that functionality is not currently implemented for irregular observations (like the space dimensions in that case). This is a work in progress in #536. You could in principle checkout that branch and try that functionality. You could also compute the coefficients "manually", as it is simply a least squares problem.

> (Currently I only represent the surface in a functional form and then do FPCA without considering the time, would considering the time bring me any advantage ?)

Well, unless I have mistaken your explanation of the data, time is the only functional measurement for each observation, as the latitude and longitude are **across** observations. Thus if you do not plan on using time, you can use any multivariate package such as scikit-learn. As for if that is advisable, I do not know, as I do not know well your data and your research problem.

**Reply by mat-ej** [reply]
Perhaps I need to read more about the functional view on data it seems. 

My objective is a regression of **y** with  **X**, (Lat, Lon) map of weather data as the feature vector. The points on map are highly spatially correlated and hence I hoped for spatial representation to be able to "compress" this data before the regression. 

My data is perfectly equispaced across space and also time, so regular structure I assume. Yes it can be 1 curve for each station, hence the data has dimension.

**X** = (Time, Latitude, Longitude)

**X** = (92472, 50, 50), y = (92490)

But there is strong time autocorrelation in time for each of measurements and also the target, (and other structures w.r.t. to time such as seasonality)

So, I might reshape the data into "repetition" of days, (or years) for each of the stations to get Time X Lat x Lon data ?
i.e. X would have the shape

**X** = (Day Count, Daily Hour, Lat, Lon)

**X** = (3853, 24, 50, 50), y = 3853

but then I assume my target y would have to be daily as well.

In any case, so far the spatial representation does improve my model where I simply represent the equispaced (Latitude, Longitude) weather data in a functional form:

```
weather_basis = TensorBasis([
    BSplineBasis(n_basis=n_basis),
    BSplineBasis(n_basis=n_basis)
])
```

This allows me to simply FPCA various weather measurements, (each separately fcourse) into very few features and use them for regression. 

I was only wondering about the time dimension If I would then be able to calculate derivatives of "spatial" weather w.r.t to time and use it, but I am still trying to understand if it would be any different from simply diffing each stations measurements with its previous observation etc. 

I found the FDAs ability to deal with missing data both in space and time very interesting and possibly useful. 

Thanks for your patience and help : ). This is somewhat new view on data for me so I am still trying to wrap my head around the concepts and see when it is beneficial and when should FDA be used. 

Have a great day.

**Reply by vnmabus** [reply]
I think that the way to frame this problem depends on which data do you have at prediction time. Consider the following examples:

**Example 1:**
At prediction time I have a functional variable (e.g. annual curve of temperatures of the station). I want to predict some scalar value (e.g. mean annual precipitation). This is a problem of scalar-on-function regression (or functional regression with scalar response). You can apply here [linear regression](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.LinearRegression.html#skfda.ml.regression.LinearRegression), [k-NN regression](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.KNeighborsRegressor.html#skfda.ml.regression.KNeighborsRegressor), [kernel regression](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.KernelRegression.html#skfda.ml.regression.KernelRegression) or [FPCA regression](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.FPCARegression.html#skfda.ml.regression.FPCARegression) (which is FPCA + Linear regression in the FPCA basis).

**Example 2:**
Same as above, but I have additional scalar variables (such as latitude or longitude) or functional variables (such as annual curve of wind speed). This is mixed data (in the first case) or multivariate functional data (in the second case), with scalar response. You can still apply [linear regression](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.LinearRegression.html#skfda.ml.regression.LinearRegression), but in order to apply other regression methods based on distances, such as k-NN, you would need to define your own distances, as we currently do not provide distances for mixed or multivariate functional data.

**Example 3:**
At prediction time I have just scalar variables (e.g. latitude and longitude). I want to predict a whole curve (e.g. the annual curve of temperatures). This is function-on-scalar regression. You can apply the distance-based methods such as [k-NN regression](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.KNeighborsRegressor.html#skfda.ml.regression.KNeighborsRegressor) or [kernel regression](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.KernelRegression.html#skfda.ml.regression.KernelRegression). The implementation of linear regression for this model is currently being done in #531, if you want to try it.

**Example 4:**
At prediction time I have scalar variables (e.g. latitude, longitude and time). The response is also scalar (e.g. temperature at that time and position). This is standard multivariate regression. This particular case can be seen as an alternative way of expressing **Example 3**. Note however that in **Example 3** we could force certain smoothness on the predicted functions, either by choosing a basis or via regularization, which is not done here.

**Example 5:**
We have at prediction time a functional variable (e.g. annual curve of temperatures of the station) and another functional variable as the response (e.g. annual curve of precipitations). This is function-on-function regression. Again you can apply distance-based methods. Concurrent linear regression (in which the point t of the response depends only on the point t of the covariate) is implemented in #531, but not yet merged. A [historical linear regression model](https://fda.readthedocs.io/en/latest/modules/ml/autosummary/skfda.ml.regression.HistoricalLinearRegression.html#skfda.ml.regression.HistoricalLinearRegression) (in which the response can depend on points s ≤ t) is already available as a separate class. The general linear case is not yet implemented.

**Example 6:**
Same as above, but with additional scalar or functional covariates. We still do not have implemented a linear regression model for this case. Distance-based methods would require your own definition of the distance.

You should consider what is your case, to see which model applies.

**Reply by mat-ej** [reply]
Great great answer, thanks so much.

**Reply by mat-ej** [reply]
So from the error messages I seem to have understood that regularization is not possible for 2D bases, is that correct ? 

```
# Combine them into a TensorBasis
basis_b = TensorBasis([lon_basis_b, lat_basis_b])

smoother = skfda.preprocessing.smoothing.BasisSmoother(
    basis=basis_b,
    regularization=regularization,
    return_basis=True,
)

fd_basis = smoother.fit_transform(fd)
```

`NotImplementedError:  basis does not support the construction of a basis of the derivatives.`

**Reply by vnmabus** [reply]
You are right, it is not currently implemented.

**Comment by mat-ej** [comment]
Thanks for answer, I ll move my question to the discussions

**Reply by vnmabus** [reply]
Moved to discussions.

**Comment by mat-ej** [comment]
So from the error messages I seem to have understood that regularization is not possible for 2D bases, 

```
# Combine them into a TensorBasis
basis_b = TensorBasis([lon_basis_b, lat_basis_b])

smoother = skfda.preprocessing.smoothing.BasisSmoother(
    basis=basis_b,
    regularization=regularization,
    return_basis=True,
)

fd_basis = smoother.fit_transform(fd)
```

`NotImplementedError:  basis does not support the construction of a basis of the derivatives.`
